source/operations.rst

```diff
@@ -0,0 +1,354 @@
+   k8s-master      Ready    master   33m   v1.15.7
+   k8s-nfs         Ready    <none>   16m   v1.15.7
+   k8s-w-ingress   Ready    <none>   16m   v1.15.7
+   egi@k8s-master:~$ helm list
+   NAME           	REVISION	UPDATED                 	STATUS  	CHART                       	APP VERSION	NAMESPACE   
+   certs-man      	2       	Wed Jan  8 15:56:58 2020	DEPLOYED	cert-manager-v0.11.0        	v0.11.0    	cert-manager
+   cluster-ingress	3       	Wed Jan  8 15:56:53 2020	DEPLOYED	nginx-ingress-1.7.0         	0.24.1     	kube-system 
+   nfs-provisioner	3       	Wed Jan  8 15:56:43 2020	DEPLOYED	nfs-client-provisioner-1.2.8	3.1.0      	kube-system 
+
+
+Modifying/Destroying the cluster
+::::::::::::::::::::::::::::::::
+
+You should be able to change the number of workers in the cluster and re-apply
+terraform to start them and then execute the playbook to get them added to the
+cluster.
+
+Any changes in the master, NFS or ingress VMs should be done carfully as those
+will probably break the configuration of the kubernetes cluster and of any
+application running on top.
+
+.. TODO: remove nodes?
+
+.. TODO: update master/ingress/nfs
+
+Destroying the cluster can be done with a single command:
+
+.. code-block:: shell
+
+   $ terraform destroy
+
+Notebooks deployments
+=====================
+
+Once the k8s cluster is up and running, you can deploy a notebooks instance.
+For each deployment you should create a file in the `deployments` directory
+following the template provided:
+
+.. code-block:: shell
+
+   $ cp deployments/hub.yaml.template deployments/hub.yaml
+   
+
+Each deployment will need a a domain name pointing to your ingress host, you
+can create one at the `FedCloud dynamic DNS service <https://nsupdate.fedcloud.eu/>`_.
+
+Then you will need to create a client for EGI Check-in to authorise users
+into the new deployment. You can create a client by going to the `Check-in demo
+OIDC clients management <https://aai-demo.egi.eu/oidc/manage/admin/clients>`_.
+Use the followin as redirect URL: ``https://<your host domain name>/hub/oauth_callback``.
+
+In the `Access` tab, add ``offline_access`` to the list of scopes. Save the
+client and take note of the client id and client secret for later.
+
+Finally you will also need 3 different random strings generated with
+``openssl rand -hex 32`` that will be used as secrets in the file describing 
+the deployment. 
+
+Go and edit the deployment description file to add this information (search for
+``# FIXME NEEDS INPUT`` in the file to quickly get there)
+
+For deploying the notebooks instance we will also use ``ansible``:
+
+.. code-block:: shell
+
+   $ ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=silently TF_STATE=./terraform ansible-playbook \
+          --inventory-file=$(which terraform-inventory) playbooks/notebooks.yaml
+   
+The first deployment trial may fail due to a timeout caused by the downloading
+of the container images needed. You can retry after a while to re-deploy.
```

Comment:
Are the ansible playbooks idempotent?

Reply:
yep :)

ansible here is mostly used as a way to avoid having access to the k8s master with public IP or sshing through a bastion. The playbook will upgrade the helm of the notebooks and that will not make any difference to what's deployed if the description of the service does not change. A new upgrade will be registered to the deployment (but I think is harmless to have the count increased +1)

Reply:
Perfect, closing that comment.

Reply:
The k8s setup playbook seems to be also idempotent, so you can add nodes easily without breaking stuff. That said, changing the master, the NFS or the ingress will cause disruptions